### PR TITLE
grandpa: make the VotingRule API async

### PR DESCRIPTION
**Pull request by andresilva**
Makes the API for `VotingRule` asynchronous and by extension changes `Environment::best_chain_containing` to return a `Future`.

Depends on https://github.com/paritytech/finality-grandpa/pull/131.

polkadot companion: https://github.com/paritytech/polkadot/pull/2422

### CODE REVIEW

**Review by rphmeier — APPROVED** [review]

File: client/finality-grandpa/src/voting_rule.rs [hunk 1]
```diff
@@ -80,37 +88,40 @@ impl<Block, B> VotingRule<Block, B> for BeforeBestBlockBy<NumberFor<Block>> wher
...
 		_base: &Block::Header,
 		best_target: &Block::Header,
 		current_target: &Block::Header,
-	) -> Option<(Block::Hash, NumberFor<Block>)> {
+	) -> VotingRuleResult<Block> {
 		use sp_arithmetic::traits::Saturating;
 
 		if current_target.number().is_zero() {
-			return None;
+			return Box::pin(std::future::ready(None));
```
Comment (bkchr) [inline client/finality-grandpa/src/voting_rule.rs:99]:
```suggestion
			return Box::pin(async { None });
```

File: client/finality-grandpa/src/voting_rule.rs [hunk 2]
```diff
@@ -60,12 +68,12 @@ impl<Block, B> VotingRule<Block, B> for () where
...
 		&self,
-		_backend: &B,
+		_backend: Arc<B>,
 		_base: &Block::Header,
 		_best_target: &Block::Header,
 		_current_target: &Block::Header,
-	) -> Option<(Block::Hash, NumberFor<Block>)> {
-		None
+	) -> VotingRuleResult<Block> {
+		Box::pin(std::future::ready(None))
```
Comment (bkchr) [inline client/finality-grandpa/src/voting_rule.rs:76]:
```suggestion
		Box::pin(async { None })
```

File: client/finality-grandpa/src/voting_rule.rs [hunk 3]
```diff
@@ -80,37 +88,40 @@ impl<Block, B> VotingRule<Block, B> for BeforeBestBlockBy<NumberFor<Block>> wher
...
+			return Box::pin(std::future::ready(None));
 		}
 
 		// find the target number restricted by this rule
 		let target_number = best_target.number().saturating_sub(self.0);
 
 		// our current target is already lower than this rule would restrict
 		if target_number >= *current_target.number() {
-			return None;
+			return Box::pin(std::future::ready(None));
```
Comment (bkchr) [inline client/finality-grandpa/src/voting_rule.rs:107]:
```suggestion
			return Box::pin(async { None });
```

File: client/finality-grandpa/src/voting_rule.rs [hunk 4]
```diff
@@ -80,37 +88,40 @@ impl<Block, B> VotingRule<Block, B> for BeforeBestBlockBy<NumberFor<Block>> wher
...
 
+		let current_target = current_target.clone();
+
 		// find the block at the given target height
-		find_target(
-			backend,
-			target_number,
-			current_target,
-		)
+		Box::pin(std::future::ready(find_target(
```
Comment (bkchr) [inline client/finality-grandpa/src/voting_rule.rs:113]:
```suggestion
		Box::pin(async move { find_target(
```

File: client/finality-grandpa/src/voting_rule.rs [hunk 5]
```diff
@@ -80,37 +88,40 @@ impl<Block, B> VotingRule<Block, B> for BeforeBestBlockBy<NumberFor<Block>> wher
...
+			&*backend,
+			target_number.clone(),
+			&current_target,
+		)))
```
Comment (bkchr) [inline client/finality-grandpa/src/voting_rule.rs:117]:
```suggestion
		)})
```

File: client/finality-grandpa/src/voting_rule.rs [hunk 6]
```diff
@@ -138,15 +149,15 @@ impl<Block, B> VotingRule<Block, B> for ThreeQuartersOfTheUnfinalizedChain where
 
 		// our current target is already lower than this rule would restrict
 		if target_number >= *current_target.number() {
-			return None;
+			return Box::pin(std::future::ready(None));
```
Comment (bkchr) [inline client/finality-grandpa/src/voting_rule.rs:152]:
```suggestion
			return Box::pin(async { None });
```

File: client/finality-grandpa/src/voting_rule.rs [hunk 7]
```diff
@@ -138,15 +149,15 @@ impl<Block, B> VotingRule<Block, B> for ThreeQuartersOfTheUnfinalizedChain where
...
 		}
 
 		// find the block at the given target height
-		find_target(
-			backend,
+		Box::pin(std::future::ready(find_target(
```
Comment (bkchr) [inline client/finality-grandpa/src/voting_rule.rs:156]:
```suggestion
		Box::pin(async move { find_target(
```

File: client/finality-grandpa/src/voting_rule.rs [hunk 8]
```diff
@@ -138,15 +149,15 @@ impl<Block, B> VotingRule<Block, B> for ThreeQuartersOfTheUnfinalizedChain where
...
+			&*backend,
 			target_number,
 			current_target,
-		)
+		)))
```
Comment (bkchr) [inline client/finality-grandpa/src/voting_rule.rs:160]:
```suggestion
		)})
```

**Review by bkchr — APPROVED** [review]

**Comment by andresilva** [comment]
@rphmeier Let me know when you test this on polkadot to make sure there aren't any shortcomings that we might need to address before merging. I'll do a release of `finality-grandpa` afterwards and remove the cargo patch.

**Comment by rphmeier** [comment]
@andresilva I got all tests passing by using this branch!

**Comment by rphmeier** [comment]
bot merge

**Pull request merged**